drivers/soundwire/slave.c

```diff
@@ -70,6 +71,17 @@ int sdw_slave_add(struct sdw_bus *bus,
 	list_add_tail(&slave->node, &bus->slaves);
 	mutex_unlock(&bus->bus_lock);
 
+	/*
+	 * The Soundwire driver probe may optionally register SDCA
+	 * sub-devices, one per Function. This means the information
+	 * on the SDCA revision and the number/type of Functions need
+	 * to be extracted from platform firmware before the SoundWire
+	 * driver probe, and as a consequence before the SoundWire
+	 * device_register() below.
+	 */
+	sdca_lookup_interface_revision(slave);
+	sdca_lookup_functions(slave);
```

Comment:
Does this really need to be called from the core, feels like it makes things a lot nicer if it is called from the end driver rather than the soundwire core?

Reply:
You missed my note in the commit message

"
Note that the context extracted in 1) should be arguably be handled
completely in the codec driver probe. That would however make it
difficult to use the ACPI information for machine quirks, and
e.g. select different machine driver and topologies as done for the
RT712_VB handling later in the series. To make the implementation of
quirks simpler, this patchset extracts a minimal amount of context
(interface revision and number/type of Functions) before the codec
driver probe, and stores this context in the scope of the 'struct
sdw_slave'.
"

Reply:
Yeah the machine match stuff is kinda interesting I kinda wonder if it could actually be handy for some of the cs42l43 quirks, but I really don't like the way it splits the disco parsing and kinda forces one to rely on a disco constant for the function type. Whilst I imagine most parts will use a disco constant for that I think it would be totally fine for that not to be a constant no?

Just to make sure I understand the situation you have with this part, the part has all the same IDs etc, but the later version of the part adds a smart mic feature? I guess I can't help but wonder if this is actually being driven by a larger issue in the machine driver, in that really with a class device one probably wants to add all the DAI links conditionally on the disco.

I am still pretty deep in a noodling session based on one of your slightly earlier pull requests, still focusing on the register map stuff. Think I have mostly got to the point where I can parse the whole register map from DisCo and handle the disco constants and fixed/default values. I will try to tidy things up a bit and port stuff onto your newer pull here. Hopefully I can then push a series for some discussion, the reason I am looking at this bit is the splitting of the disco parsing is a little tricky for the regmap stuff as it wants to see all the disco and well it feels a bit like the wrong layer to be parsing it at the soundwire layer. Also in keeping with the library concept feel nicer to have the driver call something to parse it. Will do some more thinking and report back.

Reply:
the problem statement is that the part has all the same IDs etc, but the later version of the part adds a smart mic ~feature~ function. that means we need a topology that adds the mic path.

We could add the dailink for the mic conditionally, but it's not necessary. The rule is that the machine driver needs to expose all the dailinks referenced by the topology, but it's ok to expose additional dailinks not used by the topology. So if the machine driver exposes a mic path with RT713 VA, it'd be just fine. Nothing bad would happen. In other words, exposing all links unconditionally actually makes the machine driver less complex.

Reply:
Re: the parsing stuff, I don't disagree that it should be handled in a codec driver for all disco properties,

The main issue is that the hardware discovery MUST work reliably without depending on if/when the codec driver probe happens. if the codec driver is blacklisted and modprobe'd manually 'later', we still want the information on topology to be available. That's the reason why I added this short parsing of ACPI stuff before the codec driver probe happens.